### ISSUE TITLE
Headers and signature validation

### ISSUE THREAD

**Issue by tonyg** [w3c/pubsub#14]
The draft of 20 Oct 2016 includes, in section 6.1.2, the text

> Please note however that this signature only ensures that the payload was not forged. Since the notification also includes headers, these should not be considered as safe by the subscriber, unless of course the subscriber uses HTTPS [RFC2818] callbacks.

I don't understand how HTTPS callbacks help the subscriber trust that the headers aren't forged. I imagine that most HTTPS servers (hosting the receive end of a callback) wouldn't validate the identity of a connecting client, and so would be susceptible to replay of previous payloads with altered headers.

Perhaps it would be better to amend the text to read as follows:

> Please note however that this signature only ensures that the payload was not forged. Since the notification also includes headers, these should not be considered as safe by the subscriber.

**Comment by dret** [comment]
> I don't understand how HTTPS callbacks help the subscriber trust that the headers aren't forged. I imagine that most HTTPS servers (hosting the receive end of a callback) wouldn't validate the identity of a connecting client, and so would be susceptible to replay of previous payloads with altered headers.

i had the same problem. my suggestion would have been to mention and maybe even recommend client certificates, but these are hard to implement and unlikely to be used by many.

i do agree that the language should be changed. right now it creates the wrong impression that HTTPS helps in this scenario, which it does not.

**Comment by aaronpk** [comment]
I agree, it seems to make the most sense to drop the second half of that sentence

> unless of course the subscriber uses HTTPS [RFC2818] callbacks

**Comment by dret** [comment]
> On Oct 24, 2016, at 18:08, Tony Garnock-Jones <notifications@github.com> wrote:
> I don't understand how HTTPS callbacks help the subscriber trust that the headers aren't forged. I imagine that most HTTPS servers (hosting the receive end of a callback) wouldn't validate the identity of a connecting client, and so would be susceptible to replay of previous payloads with altered headers.
> 

i had the same problem. i thought that maybe mentioning client certificates (which are optional in HTTPS) could help, but maybe that is complicating things too much.

i do agree that the text as it is does imply some level of confidence that just using HTTPS doesn't really give you.

**Comment by julien51** [comment]
My understanding of how https works is that it guarantees the integrity of the whole message, so if the signature is valid, you can assume the body has been sent by the hub, and since you could not distinguish the body from the headers in an HTTPS request, this means the whole content (body + headers) can be safely assumed to be coming from the hub.

**Comment by aaronpk** [comment]
HTTPS will ensure the integrity of the whole message, but only from the client's perspective connecting to the HTTPS server. In this case, the client is the hub, and the server is the subscriber, so there is no integrity provided to the subscriber about the hub. In other words, anybody can make a valid HTTPS request to the subscriber.

**Comment by julien51** [comment]
Of course, anybody can make a valid HTTPS request to the subscriber, but nobody can make one that includes the correct signature BUT the hub.

**Comment by aaronpk** [comment]
That's true, and HTTPS ensures the secrecy of the signature, but the sentence in question is misleading since it implies that HTTPS provides some sort of assurance that the headers are genuine, which it does not.

**Comment by julien51** [comment]
Do you mean someone could create an HTTPS request to a subscriber which would contain a body, a matching signature and other (forged) headers without knowing the secret?

**Comment by aaronpk** [comment]
I think I see what you're getting at. I think there is a way to rephrase this to make that point clearer.

I think you're saying that: 1) if the request has a valid signature, but was sent to an HTTP endpoint, it's still possible that something in between has modified the headers, so the headers should not be trusted. but, 2) if the request has a valid signature and was sent to an HTTPS endpoint, then you can trust the headers since nothing in between would have been able to modify the request. 

Does that sound right? If so, I can think of some wording to clear it up in the spec.

**Comment by julien51** [comment]
This is exactly what I mean... sorry if I was unclear initially. Any help with words is always greatly appreciated :)

**Comment by sknebel** [comment]
In addition, it might make sense to explicitly note that if the subscriber generates a token tied to the topic+hub, it can and should validate the topic+hub through use of the correct token.

**Comment by julien51** [comment]
@sandhawke I think this is at the implementation level and does not impact the interrop but indeed this is what subscribers should do for their own security.

Closing now. Feel free to re-open.

**Comment by akuckartz** [comment]
@julien51 How was this resolved?

**Comment by aaronpk** [comment]
This was resolved by removing the problematic text as suggested by @tonyg
